src/plugins/chart_expressions/common/color_categories.ts

```diff
@@ -15,27 +15,38 @@ import { isMultiFieldKey } from '@kbn/data-plugin/common';
  */
 export function getColorCategories(
   rows: DatatableRow[],
-  accessor?: string
+  accessor?: string,
+  isTransposed?: boolean,
+  exclude?: any[]
 ): Array<string | string[]> {
-  return accessor
-    ? rows.reduce<{ keys: Set<string>; categories: Array<string | string[]> }>(
-        (acc, r) => {
-          const value = r[accessor];
-          if (value === undefined) {
-            return acc;
-          }
+  const ids = isTransposed
+    ? Object.keys(rows[0]).filter((key) => accessor && key.endsWith(accessor))
+    : accessor
+    ? [accessor]
+    : [];
+
+  return rows
+    .flatMap((r) =>
+      ids
+        .map((id) => r[id])
+        .filter((v) => !(v === undefined || exclude?.includes(v)))
+        .map((v) => {
           // The categories needs to be stringified in their unformatted version.
           // We can't distinguish between a number and a string from a text input and the match should
           // work with both numeric field values and string values.
-          const key = (isMultiFieldKey(value) ? [...value.keys] : [value]).map(String);
+          const key = (isMultiFieldKey(v) ? [...v.keys] : [v]).map(String);
```

Comment:
nit: `keys` are already mapped, so no need to spread here

```suggestion
          const key = (isMultiFieldKey(v) ? v.keys : [v]).map(String);
```